### PR TITLE
Factor out object allocation from object factories [blocks: #3455]

### PR DESCRIPTION
**Pull request by danpoe**
The PR factors out the object allocation code from the Java object factory into a separate class `allocate_objectst` that can be used by both the C and Java object factories.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/allocate_objects.cpp
Comment (xbauch) [inline src/util/allocate_objects.cpp:32]: I would expect doxygen to complain about not describing the `symbols_created`.

**Review by thk123 — COMMENTED** [review]
This should probably have a full TG bump before merging (I will make after the preceding PR is merged) - also does this mean that the cprover start for C programs will do non-det init of structs like how they are handled in Java? Or have I got the wrong end of the stick? If so - very exciting!

**Comment by hannes-steffenhagen-diffblue** [comment]
@thk123 Getting something like that to work for structs and arrays is indeed the intention here

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
LGTM, I think the allocate commits should be squashed though, they don't make much sense in isolatin.

File: jbmc/src/java_bytecode/java_object_factory.cpp
Comment (hannes-steffenhagen-diffblue) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: `allocate_objectst` doesn't have an `operator()`

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4c17764420059e111371a7d4db91a205058e2470).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92382053
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9c1348b46df0796c940ea487fddfa94e6e592ae6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92725646
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by chrisr-diffblue — APPROVED** [review]
LGTM

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: Nit pick: within the folder we generally seem to use `"..."` instead of `<...>`

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: What is the input here that controls what symbol is to be created?

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: I'm not entirely happy with this terminology: the lifetime of an object vs its access scope are possibly different. In fact I don't know which one you are after for this one here. Is it about placing a symbol in the global namespace, or is it about allocation?

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: See also below: I'm used to static, stack, and heap allocation. The adjectives you use I would associate with "declaration," not "allocation." ("dynamic" is a weird one either way, but of course it so happens that it is used in CProver.)

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: I don't think you can allocate memory to an expression? Unless you are talking about the storage required by the syntax tree.

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: Ok, so "local" vs "global" is about the lifetime?

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: Not exactly the same (because it uses plain equality), but still: I'd go for `typecast_exprt::conditional_cast`.

File: src/util/allocate_objects.cpp
Comment (tautschnig) [inline src/util/allocate_objects.cpp]: As above

**Review by tautschnig — APPROVED** [review]
Overall ok, I'm just concerned about terminology.

File: src/util/allocate_objects.cpp
Comment (smowton) [reply src/util/allocate_objects.cpp]: How about "A pointer to the allocated memory will be assigned to this (lvalue) expression"?

File: src/util/allocate_objects.cpp
Comment (smowton) [reply src/util/allocate_objects.cpp]: I also think it's likely that nearly-matching types (except at the very top level) may trip later passes into introducing (e.g.) unnecessary byte-extract operations

Comment (smowton) [reply src/util/allocate_objects.cpp]: Also this (potentially over-eager) typecast omission is a change in behaviour compared to the old code AFAICT; I'd suggest in that case it would want testing with something which is `base_type_eq` but not `type_eq` (e.g. structs whose fields differ in using a symbol vs. a struct type for one of the members)

**Comment by smowton** [comment]
Unassigned for now as waiting for response re: `base_type_eq`

File: src/util/allocate_objects.cpp
Comment (danpoe) [reply src/util/allocate_objects.cpp]: Yeah this is a bit tricky terminology-wise since as you mentioned the access scope, lifetime, and the location of the allocated memory are different things.

The way I see it is that the main thing `allocate_objectst` does is generate code that allocates objects (irrespective of where the memory allocated for those objects might be located physically). The code to allocate the objects does so by either declaring a local variable (local scope, automatic lifetime), global variable (global scope, static lifetime), or using `ALLOCATE()` (dynamic).

I was thinking of renaming the members of `allocation_typet` to `AUTOMATIC_LOCAL`, `STATIC_GLOBAL`, and `DYNAMIC`. Would you be ok with that?

Comment (tautschnig) [reply src/util/allocate_objects.cpp]: Would `lifetimet` be the most suitable name (instead of `allocation_typet`)? The suggested member names sound good to me, but mostly it's the use of "type" that makes me worried (hence suggesting `lifetimet`).

Comment (danpoe) [reply src/util/allocate_objects.cpp]: Yeah, `liftetimet` sounds good to me. I'll change it to that.

File: src/util/allocate_objects.cpp
Comment (danpoe) [reply src/util/allocate_objects.cpp]: Ok, I've changed it to use `typecast_exprt::conditional_cast` (the original code used `ns.follow` on both types followed by comparing them via `==` to figure out whether a typecast should be introduced).

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 890be45178649726bc038010a4645e999285b751).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95134382
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**
